File: p1226.cpp
#include <bits/stdc++.h>

using namespace std;

class DiningPhilosophers {
private:
    mutex forks[5];
public:
    DiningPhilosophers() {
        
    }

    void wantsToEat(int philosopher,
                    function<void()> pickLeftFork,
                    function<void()> pickRightFork,
                    function<void()> eat,
                    function<void()> putLeftFork,
                    function<void()> putRightFork) {
		int left = philosopher;
        int right = (philosopher+1)%5;
        if(philosopher%2 == 1) {
            forks[left].lock();
            forks[right].lock();
        } else {
            forks[right].lock();
            forks[left].lock();
        }
        pickLeftFork();
        pickRightFork();
        eat();
        putLeftFork();
        putRightFork();
        if(philosopher%2 == 1) {            
            forks[right].unlock();
            forks[left].unlock();
        } else {            
            forks[left].unlock();
            forks[right].unlock();
        }
        
    }
};

File: p1226.py
import threading

class DiningPhilosophers:

    def __init__(self):
        self.forks = [threading.Lock() for _ in range(5)]

    def wantsToEat(self,
                   philosopher: int,
                   pickLeftFork: 'Callable[[], None]',
                   pickRightFork: 'Callable[[], None]',
                   eat: 'Callable[[], None]',
                   putLeftFork: 'Callable[[], None]',
                   putRightFork: 'Callable[[], None]') -> None:
        left = philosopher
        right = (philosopher+1)%5
        if philosopher%2 == 1:
            self.forks[left].acquire()
            self.forks[right].acquire()
        else:
            self.forks[right].acquire()
            self.forks[left].acquire()
        pickLeftFork()
        pickRightFork()
        eat()
        putLeftFork()
        putRightFork()
        if philosopher%2 == 1:
            self.forks[right].release()
            self.forks[left].release()
        else:
            self.forks[left].release()
            self.forks[right].release()
